File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DesignSystem/RolesLiveInRolesFolder.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class RolesLiveInRolesFolder: QuickSpec {
    override class func spec() {
        Given("A Role enum in an atomic component") {
            let roles = WoofHarmonize.atomicDesignPackage.enums().withSuffix("Role").withoutFolder("_Tokens")

            Then("It lives in a Roles folder next to its component") {
                roles.assertTrue(message: message) { $0.filePathString.contains("/Roles/") }
            }
        }
    }

    private static let message = LintRuleMessage(
        rule: "Component roles live in a Roles folder next to their component.",
        why: "Roles are the vocabulary a component offers (TextFontRole.bodyP1, ButtonRole.primary); keeping them next to the component makes the API discoverable.",
        howToFix: "Move the role into <Component>/Roles/.",
        badExample: "// Atoms/Text/TextColorRole.swift\nenum TextColorRole { ... }",
        goodExample: "// Atoms/Text/Roles/TextColorRole.swift\nenum TextColorRole { ... }"
    )
}

File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/designsystem/RolesLiveInRolesFolder.kt
package com.perrystreet.woof.konsist.designsystem

import com.lemonappdev.konsist.api.ext.list.withNameEndingWith
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class RolesLiveInRolesFolder : BehaviorSpec() {
    init {
        Given("A Role type in an atomic component") {
            val roles = KonsistUtils.atomicDesignModule.classes()
                .withNameEndingWith("Role")
                .filter { it.isTopLevel && !it.path.contains("/_tokens/") }

            Then("It lives in a roles package next to its component") {
                roles.assertTrue(message = Message) { it.resideInPackage("..roles") }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Component roles live in a roles package next to their component.",
            why = "Roles are the vocabulary a component offers (TextFontRole.BodyP1, ButtonRole.Primary); keeping them next to the component makes the API discoverable.",
            howToFix = "Move the role into <component>/roles/.",
            badExample = "// atoms/text/TextColorRole.kt\nenum class TextColorRole { ... }",
            goodExample = "// atoms/text/roles/TextColorRole.kt\nenum class TextColorRole { ... }",
        )
    }
}
